hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/OmSnapshotLocalDataManager.java

```diff
@@ -228,6 +307,413 @@ public void close() {
+      // Set Dirty if a version is removed.
+      setDirty();
+    }
+
+    public synchronized void commit() throws IOException {
+      // Validate modification and commit the changes.
+      SnapshotVersionsMeta localDataVersionNodes = validateModification(super.snapshotLocalData);
+      // Need to update the disk state if and only if the dirty bit is set.
+      if (isDirty()) {
+        String filePath = getSnapshotLocalPropertyYamlPath(super.snapshotId);
+        String tmpFilePath = filePath + ".tmp";
+        File tmpFile = new File(tmpFilePath);
+        boolean tmpFileExists = tmpFile.exists();
+        if (tmpFileExists) {
+          tmpFileExists = !tmpFile.delete();
+        }
+        if (tmpFileExists) {
+          throw new IOException("Unable to delete tmp file " + tmpFilePath);
+        }
+        snapshotLocalDataSerializer.save(new File(tmpFilePath), super.snapshotLocalData);
+        Files.move(tmpFile.toPath(), Paths.get(filePath), StandardCopyOption.ATOMIC_MOVE,
+            StandardCopyOption.REPLACE_EXISTING);
+        upsertNode(super.snapshotId, localDataVersionNodes);
+        // Reset dirty bit
+        resetDirty();
+      }
+    }
+
+    private void upsertNode(UUID snapshotId, SnapshotVersionsMeta snapshotVersions) throws IOException {
+      internalLock.writeLock().lock();
+      try {
+        SnapshotVersionsMeta existingSnapVersions = getVersionNodeMap().remove(snapshotId);
+        Map<Integer, LocalDataVersionNode> existingVersions = existingSnapVersions == null ? Collections.emptyMap() :
+            existingSnapVersions.getSnapshotVersions();
+        Map<Integer, List<LocalDataVersionNode>> predecessors = new HashMap<>();
+        // Track all predecessors of the existing versions and remove the node from the graph.
+        for (Map.Entry<Integer, LocalDataVersionNode> existingVersion : existingVersions.entrySet()) {
+          LocalDataVersionNode existingVersionNode = existingVersion.getValue();
+          // Create a copy of predecessors since the list of nodes returned would be a mutable set and it changes as the
+          // nodes in the graph would change.
+          predecessors.put(existingVersion.getKey(), new ArrayList<>(localDataGraph.predecessors(existingVersionNode)));
+          localDataGraph.removeNode(existingVersionNode);
+        }
+        // Add the nodes to be added in the graph and map.
+        addSnapshotVersionMeta(snapshotId, snapshotVersions);
+        // Reconnect all the predecessors for existing nodes.
+        for (Map.Entry<Integer, LocalDataVersionNode> entry : snapshotVersions.getSnapshotVersions().entrySet()) {
+          for (LocalDataVersionNode predecessor : predecessors.getOrDefault(entry.getKey(), Collections.emptyList())) {
+            localDataGraph.putEdge(predecessor, entry.getValue());
+          }
+        }
+      } finally {
+        internalLock.writeLock().unlock();
+      }
+    }
```

Comment:
Can you add a block comment explaining how this graph is supposed to work? e.g. Give an example